### PR TITLE
fix(store): add setting that allows disabling dual error handling

### PR DESCRIPTION
**Pull request by arturovt**
Issue: #1965 

NGXS has had an explicit error-handling mechanism from actions for the last years. However, this caused issues with `ErrorHandler` being called twice with the same error. The explicit error handling mechanism is required because of the NGXS action handling strategy. The default action handling process leaves the Angular zone, so NGXS actions are invoked in the `<root>` zone context. All caught actions must be returned to the Angular error handler.

Angular's default error-handling mechanism requires zone.js to be bundled. The forked Angular zone has an `onHandleError` hook which calls `zone.onError.emit(error)` when any error is being caught in Angular zone. Angular subscribes to `onError` when calls `bootstrapModuleFactory`:
```js
const exceptionHandler = moduleRef.injector.get(ErrorHandler, null);

...

const subscription = ngZone.onError.subscribe({
    next: (error) => {
        exceptionHandler.handleError(error);
    }
});
```

This PR adds "switchers" between explicit and implicit error-handling mechanisms. We either catch them manually and delegate them to `handleError` or allow Angular to catch errors in its zone and delegate them to `handleError`.

We would use the old behavior where we catch errors explicitly to be backward-compatible and allow users to switch to the implicit behavior where zone.js is responsible for catching errors and delegating them to Angular.

### CODE REVIEW

File: docs/advanced/errors.md [hunk 1]
```diff
@@ -76,39 +75,54 @@ This option allows to track errors and handle them.
...
       // error is thrown
     } catch (err) {
-      console.log('error catched inside @Action wont propagate to ErrorHandler or dispatch subscription')
+      console.log('error catched inside @Action will not propagate to ErrorHandler or dispatch subscription')
     }
   }
 ```
 
-## Handling errors after dispatching an action
+You can return an observable that completes immediately after error has been handled when dealing with streams:
```
Comment (Carniatto) [inline docs/advanced/errors.md:111]:
```suggestion
You can return an observable that completes immediately after the error has been handled when dealing with streams:
```

**Review by markwhitfeld — COMMENTED** [review]
I think that this is a great addition, but I'm unsure about the name.
What other names have you thought about?

**Comment by arturovt** [comment]
```ts
NgxsModule.forRoot([], { useNgxsErrorHandlingStrategy: false })

NgxsModule.forRoot([], { useAngularErrorHandlingStrategy: true })

NgxsModule.forRoot([], { delegateAllErrorsToErrorHandler: true })
```

**Comment by markwhitfeld** [comment]
@arturovt could you describe the two different modes?
The PR description gives a great overview and it describes the two modes as explicit and implicit error handling but the exact difference is unclear. 

Could you summarize each mode with details of behavior under each mode?

This will go a long way to help with naming.
I'll give my own attempt at this summary in the next comment.

**Comment by markwhitfeld** [comment]
Upon reviewing this, is suspect that there may actually be a fix that allows this to work as expected in all cases.
I think that there is a bug as I suspected in this comment: https://github.com/ngxs/store/pull/1927/files#r1018904814
The other bug as mentioned here (where the error is reported async):
https://github.com/ngxs/store/issues/1965#issuecomment-1396913098

could be to do with the `Promise.resolve().then(...` found here:
https://github.com/ngxs/store/pull/1927/files#diff-d211bce5ab62868a141a2954aeb1c7a4f3d3cf03dd08a8ef4f1c8496e1548f0dR32

Do you remember the motivation for this new microtask?

**Comment by arturovt** [comment]
The microtask is required because the dispatch observable may be synchronous which means the `error` callback is invoked before the `subscribed` is set to `true`.

If `useNgxsErrorHandlingStrategy` is `false`:
```
dispatch action
             |
action handler throws error
             |
     does user handle error through subscribe({ error: ... }) ?
     /                                          \
    yes                                          no
    |                                            |
propagate error to error callback    call to ErrorHandler.handleError
```

If `useNgxsErrorHandlingStrategy` is `true`:
```
dispatch action
             |
action handler throws error
             |
     does user handle error through subscribe({ error: ... }) ?
     /                                          \
    yes                                          no
    |                                            |
propagate error to error callback    call to ErrorHandler.handleError
+ call to ErrorHandler.handleError
```

**Comment by markwhitfeld** [comment]
I had a discussion with @arturovt and we closed this PR in favour of another approach that will be taken in an upcoming PR.
